Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs

```diff
@@ -348,16 +348,51 @@ protected virtual DateTimeResolutionResult ParseSpecificNight(string text, DateO
         {
             var ret = new DateTimeResolutionResult();
             var trimedText = text.Trim().ToLowerInvariant();
-            
+            var timeText = trimedText;
+
+            var match = this.Config.PeriodNightWithDateRegex.Match(trimedText);
+
+            bool hasEarly = false, hasLate = false;
+            // extract early/late prefix from text
+            if (match.Success)
+            {
+                timeText = match.Groups["night"].Value;
+                if (!string.IsNullOrEmpty(match.Groups["early"].Value))
+                {
+                    hasEarly = true;
+                    ret.Comment = "early";
+                }
+                if (!hasEarly && !string.IsNullOrEmpty(match.Groups["late"].Value))
+                {
+                    hasLate = true;
+                    ret.Comment = "late";
+                }
+            }
+
             // handle morning, afternoon..
             int beginHour, endHour, endMin = 0;
             string timeStr;
-            if (!this.Config.GetMatchedTimeRange(trimedText, out timeStr, out beginHour, out endHour, out endMin))
+            if (!this.Config.GetMatchedTimeRange(timeText, out timeStr, out beginHour, out endHour, out endMin))
             {
                 return ret;
             }
 
-            var match = this.Config.SpecificNightRegex.Match(trimedText);
+            // modify time period if "early" or "late" is existed
```

Comment:
Can you explain the logic behind this? Just from the code it's hard to figure out what exactly is happening.